tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,592 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(unsafe_code, reason = "Purpose is unsafe abstraction")]
+use super::NodeId;
+
+use std::cell::UnsafeCell;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+struct TreeNode<T> {
+    item: T,
+    children: Vec<NodeId>,
+}
+
+/// Mapping of data for the Tree Arena
+#[derive(Debug)]
+struct DataMap<T> {
+    /// The items in the tree
+    items: HashMap<NodeId, Box<UnsafeCell<TreeNode<T>>>>,
```

Comment:
This does make some sense but will have the issue that jumping to any child node from the root would require iterating through all the parents along the way (and would need to determine how removal interacts with this as would need to reindex possibly all but the node changed) - think this could also be implemented in the safe version as then parents own their children and so unsafe is not needed. 

Advantage of having the nodes stored in the hashmap is that when it is determined a node is a descendant, can use the map to jump straight to that node. 

Was also thinking whether separate parent and node hashmaps are actually needed or if it would work storing <Option<NodeId>, Box<Node>>, so only one lookup is used

Reply:
> iterating through all the parents along the way

Why? Wouldn't you be able to jump directly to the specified node's parent?

Reply:
Sorry yes, misinterpreted the suggestion - this should work I think (having the hashmap associate parents with a collection of their children) - to avoid the indexing issue on node removal, guess a HashSet or nested HashMaps could be used, tho I agree that a vec would be ideal, just need to ensure that indexing stays consistent!